pmlc/dialect/pxa/transforms/stencil.cc

```diff
@@ -0,0 +1,160 @@
+// Copyright 2019, Intel Corporation
+
+#include "llvm/ADT/Optional.h"
+#include "mlir/ADT/TypeSwitch.h"
+#include "mlir/Dialect/StandardOps/Ops.h"
+#include "mlir/IR/AffineExprVisitor.h"
+#include "mlir/Support/DebugStringHelper.h"
+#include "mlir/Dialect/AffineOps/AffineOps.h"
+
+#include "pmlc/dialect/eltwise/ir/ops.h"
+#include "pmlc/dialect/tile/ir/ops.h"
+#include "pmlc/dialect/pxa/analysis/strides.h"
+#include "pmlc/dialect/pxa/transforms/passes.h"
+#include "pmlc/dialect/pxa/ir/ops.h"
+
+#include "pmlc/util/logging.h"
+#include "pmlc/util/util.h"
+
+namespace pmlc::dialect::pxa {
+
+enum MulOperationType {
+  NoneMulOpType,
+  FloatTy,
+  IntTy,
+};
+
+struct GemmOperationMatch {
+  explicit GemmOperationMatch() {}
+
+  MulOperationType mulOpType;
+  mlir::AffineLoadOp in1Op;
+  mlir::AffineLoadOp in2Op;
+  AffineReduceOp outOp;
+};
+
+class Stencil {
+ private:
+  // Number of instructions that need to be presented in the ParallelOp region to be
+  // Considered a Gemm operation.
+  // For now they are affine.load, affined.load, mul(f/i), reduce add, terminator.
+  const unsigned kNumValidInstrInGemmRegion = 5;
+
+ public:
+  explicit Stencil() {};
+  
+  // Main function
+  void DoStenciling(mlir::AffineParallelOp op, mlir::MLIRContext* mlirContext);
+
+  // Is this a Gemm operation
+  mlir::Optional<GemmOperationMatch> getGemmOperation(mlir::AffineParallelOp op, mlir::MLIRContext* mlirContext);
+};
+
+mlir::Optional<GemmOperationMatch> Stencil::getGemmOperation(mlir::AffineParallelOp op, mlir::MLIRContext* mlirContext) {
+  mlir::Optional<GemmOperationMatch> ret;
+  Builder builder(mlirContext);
+  auto *body = op.getBody();
+  // Get the instructions in the body and match for load, load, mulXXX, reduce add operations.
+  // For everything else we fail.
+  if (body->getOperations().size() != kNumValidInstrInGemmRegion) {
+    op.emitRemark("the ParallelOp region didn't have the right number of instructions for a Gemm");
```

Comment:
Remarks like this are very helpful, but we should have a mechanism to only enable them during some level of debugging/verbosity.  We could use vlog level.

Reply:
I agree with @jbruestle  but would also add that we should probably do one or the other of IVLOG and remarks, probably not both.

Reply:
Using IVLOG only.